### PR TITLE
Infinispan extension upgrades

### LINKED ISSUE #18232: Do not produce Infinispan `RemoteCacheManager` if no Infinispan configuration is present
We depend on quarkus-infinispan-client in camel-quarkus-infinispan. Camel offers some alternative ways to configure Infinispan client, e.g. users can pass the list of hosts through the Camel route definition - see https://github.com/apache/camel-quarkus/issues/2344#issue-834501872 . If the user does not provide any `quarkus.infinispan.*` config in application.properties, starting the application fails, because quarkus-infinispan-client still produces an improperly configured `RemoteCacheManager`. Thus we'd appreciate if the producer methods in `InfinispanClientProducer` could please return `null` if no `quarkus.infinispan.*` config was provided by the end user.

cc @jamesnetherton

### PR DESCRIPTION
**Pull request by karesti**
* Upgrade to 12.1.6.Final
* ISPN-13146 Improve the documentation and defaults
* ISPN-12780 Protobuf schema registration property
* ISPN-13137 Avoid creating the Remote cache without configuration

Resolves #18232

### CODE REVIEW

**Comment by karesti** [comment]
We need to upgrade to 12.1.6.Final, don't merge please (even if all test pass which I hope they do)

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (wburns) [inline docs/src/main/asciidoc/infinispan-client.adoc:54]:
```suggestion
If you are new to Infinispan, check the 5 minute https://infinispan.org/get-started/[Getting stated with Infinispan] tutorial to
```

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (wburns) [inline docs/src/main/asciidoc/infinispan-client.adoc:57]:
```suggestion
As of Infinispan 12, authentication and authorization are enabled by default in the server, therefore you should add a user.
```

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (wburns) [inline docs/src/main/asciidoc/infinispan-client.adoc:67]:
```suggestion
Once your Infinispan Server is running, connect to the Infinispan Server with these properties.
```

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (wburns) [inline docs/src/main/asciidoc/infinispan-client.adoc:242]: The protostream module would need to be added to the project, it isn't a package in the normal protostream module.

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (wburns) [inline docs/src/main/asciidoc/infinispan-client.adoc:429]:
```suggestion
You will then need to register the Protobuf schemas
```

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (wburns) [inline docs/src/main/asciidoc/infinispan-client.adoc:431]: Should json be capitalized? I don't know :)

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (wburns) [inline docs/src/main/asciidoc/infinispan-client.adoc:435]:
```suggestion
You can check the schemas that exist under the `Schemas` tab by logging into 
the Infinispan console at  `http://localhost:11222`
```

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (wburns) [inline docs/src/main/asciidoc/infinispan-client.adoc:440]:
```suggestion
By default protobuf schemas generated this way will be registered by this extension
when the client first connects.
However, it might be required to handle the registration manually as a schema may evolve
over time when used in production, so you can disable this from occurring by configuring the 
`quarkus.infinispan-client.use-schema-registration` to `false`.
```

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (wburns) [inline docs/src/main/asciidoc/infinispan-client.adoc:442]:
```suggestion
for Kubernetes deployments, Infinispan Web Console, 
```

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (wburns) [inline docs/src/main/asciidoc/infinispan-client.adoc:441]:
```suggestion
To configure the schema manually please use https://infinispan.org/docs/infinispan-operator/master/operator.html[Infinispan Operator]
```

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (wburns) [inline docs/src/main/asciidoc/infinispan-client.adoc:445]:
```suggestion
.
```

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (wburns) [inline docs/src/main/asciidoc/infinispan-client.adoc:498]: Looks like this paragraph should be moved above with the other.

File: extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java
Comment (wburns) [inline extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java:147]: Seems this code is unneeded? It is the same as just above.

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientProducer.java
Comment (wburns) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientProducer.java:212]: This is unneeded, this is already done at https://github.com/quarkusio/quarkus/blob/main/extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientProducer.java#L167

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientProducer.java
Comment (wburns) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientProducer.java:216]: This isn't needed. Just use the config directly in my suggestion above.

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientProducer.java
Comment (wburns) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientProducer.java:68]:
```suggestion
        InfinispanClientRuntimeConfig infinispanClientRuntimeConfig = this.infinispanClientRuntimeConfig.get();
        if (infinispanClientRuntimeConfig.useSchemaRegistration.orElse(Boolean.TRUE)) {
```
There is no reason to put this in a property as isn't part of the hotrod-client.properties.

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientRuntimeConfig.java
Comment (wburns) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientRuntimeConfig.java:53]:
```suggestion
     * Enables or disables authentication. Set it to false when connecting to a Infinispan Server without authentication.
```

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientRuntimeConfig.java
Comment (wburns) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientRuntimeConfig.java:113]:
```suggestion
     * Setting this property implicitly enables SSL/TLS.
```

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientRuntimeConfig.java
Comment (wburns) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientRuntimeConfig.java:120]:
```suggestion
     * Setting this property implicitly enables SSL/TLS.
```

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientRuntimeConfig.java
Comment (wburns) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientRuntimeConfig.java:128]: Should we set the default for the config item as well?

**Review by wburns — COMMENTED** [review]
I will be on PTO for a while, so I can't follow up on a review until mid July. So it would be much appreciate if someone else can review after the changes and approve as necessary.

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (karesti) [reply docs/src/main/asciidoc/infinispan-client.adoc:431]: i capitalize

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (karesti) [reply docs/src/main/asciidoc/infinispan-client.adoc:242]: I would like to include the dependency in the client. I thought it was since I use it in the integration tests but indeed it's not in the client

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientRuntimeConfig.java
Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientRuntimeConfig.java:128]: i was doubting on this since I don't know if putting a default value here would activate the trustStore

File: extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java
Comment (karesti) [reply extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java:147]: Yes should be `SerializationContextInitializer` on the first line. On my tests using GeneratedSchema did not work and I don't want to break people still using SerializationContextInitializer

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientProducer.java
Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientProducer.java:213]: yeah I don't know what happened here, it's a mistake

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientRuntimeConfig.java
Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientRuntimeConfig.java:128]: I prefer leaving this one

**Comment by karesti** [comment]
@gsmet @geoand Will is PTO for two weeks but I addressed all this comments here. Anything else needed to merge?

File: extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java
Comment (geoand) [inline extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java:25]: Let's not use star imports please

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (geoand) [inline docs/src/main/asciidoc/infinispan-client.adoc:431]:
```suggestion
protostream provides default Protobuf
```

**Review by geoand — COMMENTED** [review]
I added a couple of tiny comments

File: extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java
Comment (karesti) [reply extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java:25]: I changed the editor config, thanks for the catch up

**Comment by geoand** [comment]
Please squash the commits when you are done :)

**Comment by karesti** [comment]
@geoand i added the changes @oraNod has done in another commit with documentation improvements

**Comment by geoand** [comment]
Cool

**Review by geoand — APPROVED** [review]

**Comment by karesti** [comment]
@geoand do we need something else to merge? will is coming back in 10 days and I addressed all his comments

**Comment by geoand** [comment]
Let's merge it, thanks!

**Comment by jamesnetherton** [comment]
Shall I create a follow up issue to #18232? It's not 100% fixed IMO.

https://github.com/quarkusio/quarkus/issues/18232#issuecomment-873922973

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (dvagnero) [inline docs/src/main/asciidoc/infinispan-client.adoc:58]: Hi @karesti , I'm checking the updates and trying to follow the steps however I got into some trouble :) 

- When I used a sample hello project with this configuration I was able to  open the console at http://127.0.1.1:11222/console/ and follow the https://infinispan.org/get-started/. There was just one little confusion between 127.0.1.1:11222 and localhost:11222. localhost:11222 didn't work.

- Then I tried the same with the quickstart and I wasn't able to compile the quickstart and failed to connect to localhost/127.0.0.1:11222. I'm probably missing some information. 
- I noticed that the configuration in the quickstart is different from the one in this guide. Which one should I use?

I have limited knowledge about Quarkus and this is my first experience with Infinispan however I wanted to share this with you in case that some of it is relevant. Please let me know if you want me to share the errors I got with the infinispan-client-quickstart. I'm on zulip or gchat.

Comment (karesti) [reply docs/src/main/asciidoc/infinispan-client.adoc:58]: The configuration in the quickstart and the guide are the same now, I did a PR to remove the unnecessary default values and simplify the connection.
 `http://localhost:11222/console/welcome` and `http://127.0.0.1:11222/console/welcome` are the same url that point to the local running infinispan server (with a container or downloading the server.
maybe you are missing a config in your `hosts` file?
thanks for trying the quickstart!

**Comment by karesti** [comment]
@jamesnetherton sorry I missed the past comment. please, create another follow up issue, I need to check with the team about protostuff and decide which is the best way here

**Comment by jamesnetherton** [comment]
Thanks @karesti, follow up issue is https://github.com/quarkusio/quarkus/issues/18517.

**Pull request merged**
